### PR TITLE
posix: implement pmemfile_[f]truncate

### PR DESCRIPTION
**Pull request by GBuella**
This change is [Reviewable](https://reviewable.io/reviews/marcinslusarz/pmemfile/55)

### CODE REVIEW

File: src/libpmemfile-core/file.c [hunk 1]
```diff
@@ -1589,3 +1589,147 @@ pmemfile_setgroups(PMEMfilepool *pfp, size_t size, const gid_t *list)
...
+		return -1;
+
+	int error = 0;
+	struct pmemfile_vinode *vinode = NULL;
+	struct pmemfile_vinode *vparent;
+	struct pmemfile_path_info info;
+	bool path_info_changed;
+
+	if (path[0] == '/')
+		vparent = pfp->root;
```
Comment (marcinslusarz) [inline src/libpmemfile-core/file.c]: You are not refing pfp->root, which means unref at the of function may free the vinode.

Also, there are 0 tests for pmemfile_truncate.

File: tests/core/rw/rw.cpp [hunk 2]
```diff
@@ -430,6 +430,95 @@ TEST_F(rw, trunc)
...
+
+	r = pmemfile_ftruncate(pfp, f2, large + 2);
+	ASSERT_EQ(r, 0) << COND_ERROR(r);
+	ASSERT_EQ(test_pmemfile_path_size(pfp, "/file2"), (large + 2));
+	ASSERT_EQ(pmemfile_lseek(pfp, f2, large, PMEMFILE_SEEK_SET), large);
+	ASSERT_EQ(pmemfile_lseek(pfp, f2, -3, PMEMFILE_SEEK_CUR), large - 3);
+	memset(buftmp, 0xff, sizeof(buftmp));
+	r = pmemfile_read(pfp, f2, buftmp, 8192);
+	ASSERT_EQ(r, 3 + 2) << COND_ERROR(r);
+	ASSERT_EQ(memcmp(buftmp, "\0\0\0testtest", 3 + 2), 0);
```
Comment (marcinslusarz) [inline tests/core/rw/rw.cpp]: "\0\0\0te"

Also, please test that <buftmp+5, buftmp+sizeof(buftmp)> contains 0xff.

File: tests/core/rw/rw.cpp [hunk 3]
```diff
@@ -430,6 +430,95 @@ TEST_F(rw, trunc)
...
+
+	r = pmemfile_ftruncate(pfp, f2, large + 11);
+	ASSERT_EQ(r, 0) << COND_ERROR(r);
+	ASSERT_EQ(test_pmemfile_path_size(pfp, "/file2"), (large + 11));
+	ASSERT_EQ(pmemfile_lseek(pfp, f2, large, PMEMFILE_SEEK_SET), large);
+	ASSERT_EQ(pmemfile_lseek(pfp, f2, -3, PMEMFILE_SEEK_CUR), large - 3);
+	memset(buftmp, 0xff, sizeof(buftmp));
+	r = pmemfile_read(pfp, f2, buftmp, 8192);
+	ASSERT_EQ(r, 3 + 11) << COND_ERROR(r);
+	ASSERT_EQ(memcmp(buftmp, "\0\0\0te\0\0\0\0\0\0\0\0\0\0", 3 + 11), 0);
```
Comment (marcinslusarz) [inline tests/core/rw/rw.cpp]: One \0 too much.

File: tests/core/rw/rw.cpp [hunk 4]
```diff
@@ -430,6 +430,95 @@ TEST_F(rw, trunc)
 
 	ASSERT_EQ(pmemfile_unlink(pfp, "/file1"), 0);
 
+	/* check ftruncate */
```
Comment (marcinslusarz) [inline tests/core/rw/rw.cpp]: Please move this block to separate test ("ftruncate"?). The missing tests for "truncate" should also be in a separate test.

**Comment by marcinslusarz** [comment]
Reviewed 3 of 5 files at r1.
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55)*

File: tests/core/rw/rw.cpp (hunk 4 above)
Comment (GBuella) [inline tests/core/rw/rw.cpp]: separate test fixture, or a now directory?

Comment (marcinslusarz) [inline tests/core/rw/rw.cpp]: fixture

Comment (GBuella) [inline tests/core/rw/rw.cpp]: Done

File: tests/core/rw/rw.cpp (hunk 3 above)
Comment (GBuella) [inline tests/core/rw/rw.cpp]: Done

File: tests/core/rw/rw.cpp (hunk 2 above)
Comment (GBuella) [inline tests/core/rw/rw.cpp]: Done

File: src/libpmemfile-core/file.c (hunk 1 above)
Comment (GBuella) [inline src/libpmemfile-core/file.c]: Done

**Comment by marcinslusarz** [comment]
Reviewed 1 of 2 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[tests/core/rw/rw.cpp, line 595 at r2](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55#-KfpnGsF_l2UpG5S1n4d:-KfpnGsF_l2UpG5S1n4e:b-7cfx6f) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/5c5aa12756739cad0ee882a5c4dd2fc4bc9a6b64/tests/core/rw/rw.cpp#L595)):*
> ```cpp
> 
> 	static const char data0[] = "testtest";
> 	static const ssize_t l0 = sizeof(data0);
> ```

Please always subtract 1 from length of the string when using it as a binary data.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55)*

**Comment by GBuella** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[tests/core/rw/rw.cpp, line 595 at r2](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55#-KfpnGsF_l2UpG5S1n4d:-KfppujOBHzXvdDarzc6:b-896fix) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/5c5aa12756739cad0ee882a5c4dd2fc4bc9a6b64/tests/core/rw/rw.cpp#L595)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please always subtract 1 from length of the string when using it as a binary data.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55)*

**Comment by marcinslusarz** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/libpmemfile-core/file.c, line 1687 at r3](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55#-Kfq0ThorPuGyq-bK1Jh:-Kfq0ThorPuGyq-bK1Ji:byd4bgq) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/3bd47d16f718991116df407209e65854013040d6/src/libpmemfile-core/file.c#L1687)):*
> ```C
> 	}
> 
> 	resolve_pathat(pfp, cred, vparent, path, &info, 0);
> ```

Now you can use resolve_pathat_full

---

*[tests/core/rw/rw.cpp, line 444 at r3](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55#-Kfq1zyeb8ZG7cZ81FES:-Kfq1zyeb8ZG7cZ81FET:b-5savx1) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/3bd47d16f718991116df407209e65854013040d6/tests/core/rw/rw.cpp#L444)):*
> ```cpp
> 	memset(bufFF, 0xff, sizeof(bufFF));
> 
> 	f = pmemfile_open(pfp, "/file1", PMEMFILE_O_CREAT | PMEMFILE_O_RDWR, 0);
> ```

I think mode should have read and write permission bits set.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55)*

**Comment by marcinslusarz** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55#-:-KfqHVy2HeQIPJu8jcJL:bnfp4nl)*

**Comment by GBuella** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/libpmemfile-core/file.c, line 1687 at r3](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55#-Kfq0ThorPuGyq-bK1Jh:-KfqHoFek7dcvKXcu2us:b-896fix) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/3bd47d16f718991116df407209e65854013040d6/src/libpmemfile-core/file.c#L1687)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Now you can use resolve_pathat_full
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/55)*

**Pull request merged**